lib/rack/request.rb

```diff
@@ -496,26 +496,35 @@ def GET
       # This method support both application/x-www-form-urlencoded and
       # multipart/form-data.
       def POST
-        if get_header(RACK_INPUT).nil?
-          raise "Missing rack.input"
-        elsif get_header(RACK_REQUEST_FORM_INPUT) == get_header(RACK_INPUT)
-          get_header(RACK_REQUEST_FORM_HASH)
-        elsif form_data? || parseable_data?
-          unless set_header(RACK_REQUEST_FORM_HASH, parse_multipart)
-            form_vars = get_header(RACK_INPUT).read
-
-            # Fix for Safari Ajax postings that always append \0
-            # form_vars.sub!(/\0\z/, '') # performance replacement:
-            form_vars.slice!(-1) if form_vars.end_with?("\0")
-
-            set_header RACK_REQUEST_FORM_VARS, form_vars
-            set_header RACK_REQUEST_FORM_HASH, parse_query(form_vars, '&')
+        if error = get_header(RACK_REQUEST_FORM_ERROR)
+          raise error
+        end
+
+        begin
+          if get_header(RACK_INPUT).nil?
+            raise "Missing rack.input"
+          elsif get_header(RACK_REQUEST_FORM_INPUT) == get_header(RACK_INPUT)
+            get_header(RACK_REQUEST_FORM_HASH)
+          elsif form_data? || parseable_data?
+            unless set_header(RACK_REQUEST_FORM_HASH, parse_multipart)
+              form_vars = get_header(RACK_INPUT).read
+
+              # Fix for Safari Ajax postings that always append \0
+              # form_vars.sub!(/\0\z/, '') # performance replacement:
+              form_vars.slice!(-1) if form_vars.end_with?("\0")
+
+              set_header RACK_REQUEST_FORM_VARS, form_vars
+              set_header RACK_REQUEST_FORM_HASH, parse_query(form_vars, '&')
+            end
+            set_header RACK_REQUEST_FORM_INPUT, get_header(RACK_INPUT)
+            get_header RACK_REQUEST_FORM_HASH
+          else
+            set_header RACK_REQUEST_FORM_INPUT, get_header(RACK_INPUT)
+            set_header(RACK_REQUEST_FORM_HASH, {})
           end
-          set_header RACK_REQUEST_FORM_INPUT, get_header(RACK_INPUT)
-          get_header RACK_REQUEST_FORM_HASH
-        else
-          set_header RACK_REQUEST_FORM_INPUT, get_header(RACK_INPUT)
-          set_header(RACK_REQUEST_FORM_HASH, {})
+        rescue => error
+          set_header(RACK_REQUEST_FORM_ERROR, error)
```

Comment:
I don't think storing an Exception instance is ideal, because it means we also cache the backtrace:

```ruby
@error = nil

def params
  if @error
    raise @error
  end

  begin
    raise "oops"
  rescue => e
    @error = e
    raise
  end
end

def method_1
  params
end

def method_2
  params
end


method_1 rescue nil
method_2
```

```
/tmp/foo.rb:9:in `params': oops (RuntimeError)
	from /tmp/foo.rb:17:in `method_1'
	from /tmp/foo.rb:25:in `<main>'
```

Note how the backtrace say `method_1` when it should be `method_2`.

I see multiple solutions:

  - Store the error class and message only
  - Raise using `raise error.class, error.message, cause: error`

Reply:
I don't have a strong opinion but I slightly prefer the latter, because it's still only one thing to store. So let's do that?

Reply:
I agree with @casperisfine that caching the error instance is bad as we don't want to keep the backtrace (and potentially nested exception references).  If you only want to store one thing:

```ruby
# When storing error
set_header(RACK_REQUEST_FORM_ERROR, [error.class, error.message])

# When raising error
raise *error
```